server/src/main/java/org/apache/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -184,12 +187,19 @@ private static class NodeRoleWatcher implements DruidNodeDiscovery, Closeable
       this.curatorFramework = curatorFramework;
       this.nodeRole = nodeRole;
       this.jsonMapper = jsonMapper;
-      this.baseNodeRoleWatcher = new BaseNodeRoleWatcher(listenerExecutor, nodeRole);
+      this.listenerExecutor = listenerExecutor;
+      this.basePath = basePath;
 
       // This is required to be single threaded from docs in PathChildrenCache.
       this.cacheExecutor = Execs.singleThreaded(
           StringUtils.format("NodeRoleWatcher[%s]", StringUtils.encodeForFormat(nodeRole.toString()))
       );
+      refreshCache();
+    }
+
+    public void refreshCache()
+    {
+      baseNodeRoleWatcher = new BaseNodeRoleWatcher(listenerExecutor, nodeRole);
```

Comment:
We should reuse this since a listener might have already been registered.

Reply:
Done